### DISCUSSION TITLE
Planned JS SDK refactoring

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#3134] [Announcements]
_I've decided post it here for better visibility._

The planned next JS SDK release comes with several improvements but unfortunately some of them would introduce breaking changes.

For most users probably that will not be the case since there are several function overloads and data normalization happening behind the scenes but please consider reviewing the below `JS SDK v0.17.0-rc1` CHANGELOG.

There are still some minor details that need to be polished but the planned changes are in the [`develop` branch](https://github.com/pocketbase/js-sdk/tree/develop) and pushed in npm as [`v0.17.0-rc1` release](https://www.npmjs.com/package/pocketbase/v/0.17.0-rc1) (_under the `next` tag_).

I'll wait for ~2 weeks for feedback in case someone come up with some other suggestion or disapproval of the changes.

---

## JS SDK 0.17.0-rc1

- To simplify file uploads, we now allow sending the `multipart/form-data` request body also as plain object if at least one of the object props has `File` or `Blob` value.
    ```js
    // the standard way to create multipart/form-data body
    const data = new FormData();
    data.set("title", "lorem ipsum...")
    data.set("document", new File(...))

    // this is the same as above
    // (it will be converted behind the scenes to FormData)
    const data = {
      "title":    "lorem ipsum...",
      "document": new File(...),
    };

    await pb.collection("example").create(data);
    ```

- Added new `pb.authStore.isAdmin` and `pb.authStore.isAuthRecord` helpers to check the type of the current auth state.

- The default `LocalAuthStore` now listen to the browser [storage event](https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event), so that we can sync automatically the `pb.authStore` state between multiple tabs.

- Added new helper `AsyncAuthStore` class that can be used to integrate with any 3rd party async storage implementation (_usually this is needed when working with React Native_):
    ```js
    import AsyncStorage from "@react-native-async-storage/async-storage";
    import PocketBase, { AsyncAuthStore } from "pocketbase";

    const store = new AsyncAuthStore({
        save:    async (serialized) => AsyncStorage.setItem("pb_auth", serialized),
        initial: await AsyncStorage.getItem("pb_auth"),
    });

    const pb = new PocketBase("https://example.com", store)
    ```

- ⚠️ All API actions now return plain object (POJO) as response, aka. the custom class wrapping was removed and you no longer need to manually call `structuredClone(response)` when using with SSR frameworks.

    This could be a breaking change if you use the below classes (_and respectively their helper methods like `$isNew`, `$load()`, etc._) since they were replaced with plain TS interfaces:
    ```
    class BaseModel    -> interface BaseModel
    class Admin        -> interface AdminModel
    class Record       -> interface RecordModel
    class LogRequest   -> interface LogRequestModel
    class ExternalAuth -> interface ExternalAuthModel
    class Collection   -> interface CollectionModel
    class SchemaField  -> interface SchemaField
    class ListResult   -> interface ListResult
    ```
    _Side-note:_ If you use somewhere in your code the `Record` and `Admin` classes to determine the type of your `pb.authStore.model`, you can safely replace it with the new `pb.authStore.isAdmin` and `pb.authStore.isAuthRecord` getters.

- ⚠️ Added support for per-request `fetch` options, including also specifying completely custom `fetch` implementation.

    In addition to the default [`fetch` options](https://developer.mozilla.org/en-US/docs/Web/API/fetch#options), the following configurable fields are supported:

    ```ts
    interface SendOptions extends RequestInit {
        // any other custom key will be merged with the query parameters
        // for backward compatibility and to minimize the verbosity
        [key: string]: any;

        // optional custom fetch function to use for sending the request
        fetch?: (url: RequestInfo | URL, config?: RequestInit) => Promise<Response>;

        // custom headers to send with the requests
        headers?: { [key: string]: string };

        // the body of the request (serialized automatically for json requests)
        body?: any;

        // query params that will be appended to the request url
        query?: { [key: string]: any };

        // the request identifier that can be used to cancel pending requests
        requestKey?:  string|null;

        // @deprecated use `requestKey:string` instead
        $cancelKey?:  string;

        // @deprecated use `requestKey:null` instead
        $autoCancel?: boolean;
    }
    ```

    Methods that supports additional optional body parameters will still continue to work but with a warning message in the console to update to the new format.
    For example:
    ```js
    // OLD (should still work but with a warning in the console)
    await pb.collection("example").authRefresh({}, {
      "expand": "someRelField",
    })

    // NEW
    await pb.collection("example").authRefresh({
      "expand": "someRelField",
      // send some additional header
      "headers": {
        "X-Custom-Header": "123",
      },
      "cache": "no-store" // also usually used by frameworks like Next.js
    })
    ```

- Internal refactoring (updated dev dependencies, refactored the tests to use Vitest instead of Mocha, etc.).

**Comment by lbensaad** [comment]
That is very good, many thanks. 
Currently I am using `v0.17.0-rc1` and the good news is that I did not have to change anything in my code. Other broken things like reactivity in Qwik started to work now. It solved a big issue for me BuilderIO/qwik#4974, especially after you replaced the Record class with Interface.
Do you have any idea when v0.17.0 will be released?

**Reply by ganigeorgiev** [reply]
If nothing critical show up I think it can be released by the end of the week.

**Reply by ganigeorgiev** [reply]
_Just for info, the above changes were published as [v0.17.0 release](https://github.com/pocketbase/js-sdk/releases/tag/v0.17.0)_

**Reply by lbensaad** [reply]
Great.  Thank you.

**Comment by xxyuze** [comment]
Can a custom request implementation be globally replaced like storage in the same way?

```javascript
const pb = new PocketBase("https://example.com", store, customFetch);
```

**Reply by ganigeorgiev** [reply]
It should be possible if you register the [`beforeSend` hook](https://github.com/pocketbase/js-sdk#send-hooks). Something like:

```js
pb.beforeSend = function (url, options) {
    options.fetch = yourCustomFetchFunc;

    return { url, options };
};
```

**Reply by xxyuze** [reply]
Thanks a lot.

**Reply by xxyuze** [reply]
Some runtime environments do not support AbortController, such as the WeChat Mini Program environment in China. Although I have implemented my own fetch, I still encountered the following error. Is there a solution?

```javascript
ReferenceError: AbortController is not defined
```

**Reply by ganigeorgiev** [reply]
Environments that doesn't support `AbortControl` can load a polyfill (you can try to search in npm) and/or define a custom global implementation (eg. `global.AbortControl = something`).

**Reply by xxyuze** [reply]
It works, Thanks!

**Reply by Saad5400** [reply]
A bit late, but I found this: https://www.npmjs.com/package/abortcontroller-polyfill

Now where should I import it or do the `global.AbortControl  = AbortControl  ` ?
